!WIP; feat: Config flag error ignored during login #251 #258

qcserestipy opened the pull request:
Fixes #251 

**This pull request is currently work in progress.**

**Description**
This pull request focuses on configuration handling and the login process.

**Changes**
- config flag is now properly handled in the runLogin function. Originally the default path was always taken.
- Added ConfigPath to loginview
- Added ValidateConfigPath function to ensure the configuration file path ends with .yaml or .yml.
- Existing config files only get updated if credentials do not yet exist

**Open Changes**
- Adding support for environment variable of config path
- Overwrite config entry when password has changed for same name

Even though this PR is still WIP I would be happy to receive your feedback on any code changes.

bupd reviewed (changes requested):
Thanks for your contribution.

Also we can store user-defined config path in data file. to persist across sessions.

- update config default location to `$XDG_CONFIG_HOME/harbor-cli/config.yaml`
- store the data file in `$XDG_DATA_HOME/harbor-cli/data` which is also `~/.local/share/harbor-cli/data`. In the data file we can store the user defined config path.

- With this we will be having the robust config option for the harbor-cli.

bupd commented:
@qcserestipy Please sign your commits

qcserestipy commented:
@bupd Thank you for your feedback! Yes, you were right. The config was set as a global flag already. I removed the option from the login screen. The credential handling is now moved to the runLogin function only. I will start looking into the $XDG_CONFIG_HOME config handling you suggested.

qcserestipy commented:
The changes ensure that the application consistently references the active configuration, manages multiple credential sets efficiently, and maintains a reliable state across different operations.

## Key Changes

1. **PersistentPreRun Function for Configuration Initialization**
   - **Purpose:** Ensures that the configuration is initialized before any command execution.
   - **Benefits:** Guarantees that all commands have access to the necessary configuration data, preventing runtime errors related to uninitialized configurations.

2. **Data File Logic for Tracking Active Configuration**
   - **Purpose:** Implements a data file (`data.yaml`) that keeps track of the last active configuration file (`config.yaml`).
   - **Benefits:** Allows users to switch between multiple configuration files seamlessly without manually specifying the config path each time.

3. **Global Pointers for DataFile and ConfigFile**
   - **Purpose:** Makes `HarborData` and `HarborConfig` accessible globally through pointers.
   - **Benefits:** Simplifies access to configuration and data throughout the application, reducing the need for passing configuration objects between functions and packages.

4. **Harbor Client for Credential Retrieval**
   - **Purpose:** Adjusts the Harbor client to retrieve credentials using the global configuration pointer.
   - **Benefits:** Ensures that all client operations consistently use the currently active credentials, enhancing security and reliability.

5. **Credential Management Enhancements in Login Command**
   - **Purpose:** Adds functionality to the login process to update credentials in the configuration file after a successful login.
   - **Benefits:** Automatically manages and updates credential sets, reducing manual intervention and minimizing the risk of credential mismatches.

## Showcase 
![bitmap](https://github.com/user-attachments/assets/f42a292c-5e78-47f2-8eac-63417a7fa352)

### 1
The datafile is created and it contains info about the config path which is in the config2.yaml file

### 2 and 3
I log in to the server and choose a new config file called config.yaml. The login proceeds, the file is created and the data file is updated.

### 4 
The data file contains information about an old file config3.yaml that does not exist anymore. The cli crashes. The user is informed to create a new blank config3.yaml or change the name to a new config file where the data file gets updated.

### 5
The usual harbor subcommands can read the peristent config file specified in data.yaml, or use a config file that is given on the fly by the --config flag.

qcserestipy commented:
Is there a good way to test this code in unit tests?

qcserestipy commented on pkg/utils/config.go:
```diff
@@ -16,87 +19,386 @@ type Credential struct {
+		log.Fatalf("Unable to determine user home directory: %v", err)
+	}
+
+	xdgConfigHome := os.Getenv("XDG_CONFIG_HOME")
```
I followed the instructions in [this specification](https://specifications.freedesktop.org/basedir-spec/latest/). According to section 3 the xdg home should default to $HOME/.config in case the XDG environment variables are not set in the shell/env. Therefore I thought it could be helpful to check whether a user has a $HOME, since this might not be fulfilled for e.g. machine users.

qcserestipy replied:
But I will double check whether the result in the expected behavior, and want to include this in some unit testing.

Vad1mo commented:
@qcserestipy make sure you comple the DCO part -> https://github.com/goharbor/harbor-cli/pull/259/checks?check_run_id=33351002182

Vad1mo commented:
@qcserestipy can you also address the discussion points we can than move forward

qcserestipy commented:
I removed the WIP tag for now

bupd reviewed (approved):
LGTM! 🚀
